File: Backend/app/models/employee.py
"""Pydantic models for HR employees and departments."""

from __future__ import annotations

from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, EmailStr, Field

EmployeeStatus = Literal[
    "active",
    "on_leave",
    "probation",
    "onboarding",
    "offboarding",
    "terminated",
]
EmploymentType = Literal["full_time", "part_time", "contract", "intern"]
OnboardingStage = Literal[
    "paperwork", "it_setup", "orientation", "buddy_assigned", "complete"
]
OffboardingStage = Literal[
    "notice",
    "knowledge_transfer",
    "asset_return",
    "exit_interview",
    "complete",
]


class EmployeeAddress(BaseModel):
    line1: str = ""
    line2: Optional[str] = None
    city: str = ""
    state: Optional[str] = None
    postalCode: Optional[str] = None
    country: str = ""


class EmployeeSkill(BaseModel):
    name: str
    level: Optional[Literal["beginner", "intermediate", "advanced", "expert"]] = None


class EmployeeCreate(BaseModel):
    firstName: str = Field(..., min_length=1, max_length=100)
    lastName: str = Field(..., min_length=1, max_length=100)
    email: EmailStr
    phone: str = ""
    jobTitle: str = Field(..., min_length=1, max_length=200)
    departmentId: Optional[str] = None
    managerId: Optional[str] = None
    location: str = "Nairobi"
    status: EmployeeStatus = "onboarding"
    employmentType: EmploymentType = "full_time"
    startDate: str = Field(..., min_length=8, max_length=32)
    endDate: Optional[str] = None
    workEmail: Optional[EmailStr] = None
    personalEmail: Optional[str] = None
    dateOfBirth: Optional[str] = None
    nationality: Optional[str] = None
    displayName: Optional[str] = None
    pronouns: Optional[str] = None
    gender: Optional[str] = None
    maritalStatus: Optional[str] = None
    jobGrade: Optional[str] = None
    probationMonths: Optional[int] = None
    workArrangement: Optional[str] = None
    weeklyHours: Optional[float] = None
    address: Optional[EmployeeAddress] = None
    emergencyContact: Optional[Dict[str, Any]] = None
    skills: List[EmployeeSkill] = Field(default_factory=list)
    tags: List[str] = Field(default_factory=list)
    leaveBalanceDays: float = 0
    performanceRating: float = 0
    baseSalary: float = 0
    currency: str = "USD"
    totalCompensation: Optional[float] = None
    equityValue: float = 0
    bonusTarget: Optional[float] = None
    signOnBonus: Optional[float] = None
    bankDetails: Optional[Dict[str, Any]] = None
    onboardingChecklist: Optional[Dict[str, Any]] = None
    invitePreferences: Optional[Dict[str, Any]] = None
    onboardingStage: Optional[OnboardingStage] = None
    offboardingStage: Optional[OffboardingStage] = None
    avatarUrl: Optional[str] = None


class EmployeeUpdate(BaseModel):
    firstName: Optional[str] = Field(None, min_length=1, max_length=100)
    lastName: Optional[str] = Field(None, min_length=1, max_length=100)
    email: Optional[EmailStr] = None
    phone: Optional[str] = None
    jobTitle: Optional[str] = Field(None, min_length=1, max_length=200)
    departmentId: Optional[str] = None
    managerId: Optional[str] = None
    location: Optional[str] = None
    status: Optional[EmployeeStatus] = None
    employmentType: Optional[EmploymentType] = None
    startDate: Optional[str] = None
    endDate: Optional[str] = None
    workEmail: Optional[EmailStr] = None
    personalEmail: Optional[str] = None
    dateOfBirth: Optional[str] = None
    nationality: Optional[str] = None
    displayName: Optional[str] = None
    pronouns: Optional[str] = None
    gender: Optional[str] = None
    maritalStatus: Optional[str] = None
    jobGrade: Optional[str] = None
    probationMonths: Optional[int] = None
    workArrangement: Optional[str] = None
    weeklyHours: Optional[float] = None
    address: Optional[EmployeeAddress] = None
    emergencyContact: Optional[Dict[str, Any]] = None
    skills: Optional[List[EmployeeSkill]] = None
    tags: Optional[List[str]] = None
    leaveBalanceDays: Optional[float] = None
    performanceRating: Optional[float] = None
    baseSalary: Optional[float] = None
    currency: Optional[str] = None
    totalCompensation: Optional[float] = None
    equityValue: Optional[float] = None
    bonusTarget: Optional[float] = None
    signOnBonus: Optional[float] = None
    bankDetails: Optional[Dict[str, Any]] = None
    onboardingChecklist: Optional[Dict[str, Any]] = None
    invitePreferences: Optional[Dict[str, Any]] = None
    compensationBreakdown: Optional[List[Dict[str, Any]]] = None
    compensationHistory: Optional[List[Dict[str, Any]]] = None
    equityGrants: Optional[List[Dict[str, Any]]] = None
    attendanceSummary: Optional[Dict[str, Any]] = None
    attendanceTrend: Optional[Dict[str, Any]] = None
    timesheet: Optional[List[Dict[str, Any]]] = None
    leaveHistory: Optional[List[Dict[str, Any]]] = None
    goals: Optional[List[Dict[str, Any]]] = None
    peerFeedback: Optional[List[Dict[str, Any]]] = None
    ratingHistory: Optional[Dict[str, Any]] = None
    documents: Optional[List[Dict[str, Any]]] = None
    activity: Optional[List[Dict[str, Any]]] = None
    onboardingStage: Optional[OnboardingStage] = None
    offboardingStage: Optional[OffboardingStage] = None
    avatarUrl: Optional[str] = None
    tenureYears: Optional[float] = None


EmployeeDocumentCategory = Literal[
    "national_id",
    "good_conduct",
    "cv",
    "tax_id",
    "contract",
    "certificate",
    "other",
]


class EmployeeDocumentCreate(BaseModel):
    """Metadata for a file already uploaded via /api/upload."""

    name: str = Field(..., min_length=1, max_length=200)
    category: EmployeeDocumentCategory = "other"
    fileUrl: str = Field(..., min_length=1, max_length=2000)
    fileName: str = Field(..., min_length=1, max_length=500)
    fileSize: int = Field(0, ge=0)


class EmployeeSelfServiceUpdate(BaseModel):
    """Fields an employee may update on their own roster profile."""

    firstName: Optional[str] = Field(None, min_length=1, max_length=100)
    lastName: Optional[str] = Field(None, min_length=1, max_length=100)
    displayName: Optional[str] = Field(None, max_length=200)
    phone: Optional[str] = Field(None, max_length=40)
    personalEmail: Optional[str] = Field(None, max_length=254)
    location: Optional[str] = Field(None, max_length=200)
    dateOfBirth: Optional[str] = Field(None, max_length=32)
    nationality: Optional[str] = Field(None, max_length=100)
    pronouns: Optional[str] = Field(None, max_length=40)
    gender: Optional[str] = Field(None, max_length=40)
    maritalStatus: Optional[str] = Field(None, max_length=40)
    address: Optional[EmployeeAddress] = None
    emergencyContact: Optional[Dict[str, Any]] = None
    avatarUrl: Optional[str] = Field(None, max_length=2000)


class EmployeeLeaveRequestCreate(BaseModel):
    """Employee self-service leave request payload."""

    leaveTypeId: str = Field(..., min_length=1, max_length=64)
    startDate: str = Field(..., min_length=10, max_length=32)
    endDate: str = Field(..., min_length=10, max_length=32)
    startPeriod: Optional[str] = Field("morning", max_length=32)
    endPeriod: Optional[str] = Field("end_of_day", max_length=32)
    reason: Optional[str] = Field(None, max_length=1000)
    substituteEmployeeId: Optional[str] = Field(None, max_length=64)


class EmployeeLeaveCommentCreate(BaseModel):
    """Comment on the employee's own leave request."""

    body: str = Field(..., min_length=1, max_length=2000)


class DepartmentCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=120)
    code: str = Field(..., min_length=1, max_length=32)
    headEmployeeId: Optional[str] = None
    description: Optional[str] = Field(None, max_length=500)
    parentId: Optional[str] = None


class DepartmentUpdate(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=120)
    code: Optional[str] = Field(None, min_length=1, max_length=32)
    headEmployeeId: Optional[str] = None
    description: Optional[str] = Field(None, max_length=500)
    parentId: Optional[str] = None


class EmployeeImportRow(BaseModel):
    firstName: str = Field(..., min_length=1, max_length=100)
    lastName: str = Field(..., min_length=1, max_length=100)
    email: EmailStr
    jobTitle: str = Field(..., min_length=1, max_length=200)
    phone: Optional[str] = None
    location: Optional[str] = None
    status: Optional[EmployeeStatus] = None
    employmentType: Optional[EmploymentType] = None
    startDate: Optional[str] = None
    workEmail: Optional[EmailStr] = None
    departmentCode: Optional[str] = None
    departmentName: Optional[str] = None


class EmployeeImportRequest(BaseModel):
    # Pydantic v1 enforces list size via min_items/max_items (not min_length/max_length).
    rows: List[EmployeeImportRow] = Field(..., min_items=1, max_items=500)

File: types/employee.ts
/** HR employee types for the admin People module (/api/admin/employees). */

export type EmployeeStatus =
  | "active"
  | "on_leave"
  | "probation"
  | "onboarding"
  | "offboarding"
  | "terminated";

export type EmploymentType = "full_time" | "part_time" | "contract" | "intern";

export type OnboardingStage =
  | "paperwork"
  | "it_setup"
  | "orientation"
  | "buddy_assigned"
  | "complete";

export type OffboardingStage =
  | "notice"
  | "knowledge_transfer"
  | "asset_return"
  | "exit_interview"
  | "complete";

export interface EmployeeAddress {
  line1: string;
  line2?: string;
  city: string;
  state?: string;
  postalCode?: string;
  country: string;
}

export interface EmployeeSkill {
  name: string;
  level?: "beginner" | "intermediate" | "advanced" | "expert";
}

export interface CompensationBreakdown {
  label: string;
  amount: number;
  color: string;
}

export interface CompensationRecord {
  effectiveDate: string;
  type: "raise" | "promotion" | "adjustment" | "bonus" | "equity_grant";
  previousBase?: number;
  newBase?: number;
  note: string;
}

export interface EquityGrant {
  id: string;
  grantDate: string;
  shares: number;
  vestedShares: number;
  vestSchedule: string;
  cliffMonths: number;
  status: "active" | "fully_vested" | "cancelled";
  grantType?: "rsu" | "options";
  grantValue?: number;
}

export interface AttendanceDay {
  date: string;
  status: "present" | "remote" | "leave" | "late" | "absent" | "holiday";
  checkIn?: string;
  checkOut?: string;
  overtimeHours?: number;
}

export interface LeaveRecord {
  id: string;
  type: string;
  startDate: string;
  endDate: string;
  days: number;
  status: "approved" | "pending" | "rejected";
}

export interface PerformanceGoal {
  id: string;
  title: string;
  progress: number;
  dueDate: string;
  status: "on_track" | "at_risk" | "completed" | "not_started";
}

export interface PeerFeedback {
  id: string;
  fromName: string;
  date: string;
  summary: string;
  rating: number;
}

export type EmployeeDocumentCategory =
  | "national_id"
  | "good_conduct"
  | "cv"
  | "tax_id"
  | "contract"
  | "certificate"
  | "other";

export type EmployeeDocumentStatus = "uploaded" | "pending" | "failed" | "sent";

export interface EmployeeDocument {
  id: string;
  name: string;
  category: EmployeeDocumentCategory;
  uploadedAt: string;
  sizeKb: number;
  url?: string;
  fileName?: string;
  status?: EmployeeDocumentStatus;
}

export const EMPLOYEE_DOCUMENT_TYPE_OPTIONS: {
  id: EmployeeDocumentCategory;
  label: string;
}[] = [
  { id: "national_id", label: "National ID" },
  { id: "good_conduct", label: "Good Conduct" },
  { id: "cv", label: "CV / Resume" },
  { id: "tax_id", label: "Tax ID / PAN" },
  { id: "contract", label: "Contract" },
  { id: "certificate", label: "Certificate" },
  { id: "other", label: "Other" },
];

export const EMPLOYEE_DOCUMENT_CATEGORY_LABELS: Record<
  EmployeeDocumentCategory,
  string
> = {
  national_id: "National ID",
  good_conduct: "Good Conduct",
  cv: "CV / Resume",
  tax_id: "Tax ID / PAN",
  contract: "Contracts",
  certificate: "Certificates",
  other: "Other",
};

export interface ActivityEntry {
  id: string;
  date: string;
  actor: string;
  action: string;
  detail?: string;
}

export interface Employee {
  id: string;
  employeeNumber: string;
  firstName: string;
  lastName: string;
  email: string;
  phone: string;
  avatarUrl?: string;
  jobTitle: string;
  departmentId: string;
  departmentName: string;
  managerId?: string;
  managerName?: string;
  location: string;
  status: EmployeeStatus;
  employmentType: EmploymentType;
  startDate: string;
  endDate?: string;
  workEmail: string;
  personalEmail?: string;
  dateOfBirth?: string;
  nationality?: string;
  displayName?: string;
  pronouns?: string;
  gender?: string;
  maritalStatus?: string;
  jobGrade?: string;
  probationMonths?: number | null;
  workArrangement?: string;
  weeklyHours?: number | null;
  address?: EmployeeAddress;
  emergencyContact?: { name: string; relation: string; phone: string };
  skills: EmployeeSkill[];
  tags: string[];
  /** Years of tenure derived for display; also computable from startDate */
  tenureYears: number;
  leaveBalanceDays: number;
  performanceRating: number;
  baseSalary: number;
  currency: string;
  totalCompensation: number;
  equityValue: number;
  bonusTarget?: number;
  signOnBonus?: number;
  bankDetails?: {
    accountName?: string;
    accountNumber?: string;
    bankName?: string;
    branch?: string;
  } | null;
  onboardingChecklist?: {
    docusignOffer?: boolean;
    sso?: boolean;
    payroll?: boolean;
    buddy?: boolean;
    equipment?: boolean;
  } | null;
  invitePreferences?: {
    method?: string;
    timing?: string;
    welcomeMessage?: string;
    sendInvite?: boolean;
    inviteStatus?: "sent" | "failed" | "skipped" | "deferred" | "sending" | string;
    pendingInvite?: boolean;
    inviteSentAt?: string | null;
    scheduledInviteAt?: string | null;
  } | null;
  compensationBreakdown: CompensationBreakdown[];
  compensationHistory: CompensationRecord[];
  equityGrants: EquityGrant[];
  attendanceSummary: {
    presentDays: number;
    leaveDays: number;
    lateDays: number;
    overtimeHours: number;
  };
  attendanceTrend: { labels: string[]; values: number[] };
  timesheet: AttendanceDay[];
  leaveHistory: LeaveRecord[];
  goals: PerformanceGoal[];
  peerFeedback: PeerFeedback[];
  ratingHistory: { labels: string[]; values: number[] };
  documents: EmployeeDocument[];
  activity: ActivityEntry[];
  onboardingStage?: OnboardingStage;
  offboardingStage?: OffboardingStage;
}

export interface Department {
  id: string;
  name: string;
  code: string;
  headEmployeeId?: string;
  headName?: string;
  employeeCount: number;
  description?: string;
  parentId?: string;
  createdAt?: string;
  updatedAt?: string;
}

export interface EmployeeListFilters {
  q?: string;
  departmentId?: string;
  status?: EmployeeStatus | "all";
  employmentType?: EmploymentType | "all";
}

export const EMPLOYEE_STATUS_LABELS: Record<EmployeeStatus, string> = {
  active: "Active",
  on_leave: "On leave",
  probation: "Probation",
  onboarding: "Onboarding",
  offboarding: "Offboarding",
  terminated: "Terminated",
};

export const EMPLOYMENT_TYPE_LABELS: Record<EmploymentType, string> = {
  full_time: "Full-time",
  part_time: "Part-time",
  contract: "Contract",
  intern: "Intern",
};
